text/subsingleton-elimination.md

```diff
@@ -0,0 +1,216 @@
+(* Elimination of an inductive object of sort Prop is not allowed on a predicate in sort Set *)
+```
+which would actually be easy to address (see point 2 discussed in item 2 of the next section).
+
+## `SProp`
+
+### `SProp`-truncation
+Similarly, let's call `SProp`-truncation the following type:
+```
+Variant SSquash (A:Type) : SProp := ssquash : A -> SSquash A.
+```
+whose content can be `match`-eliminated only in a context of type `SProp`.
+
+Then, any inductive or coinductive type declared in `SProp` is intended to be equivalent to the `SSquash` encapsulation of its declaration in `Type`.
+
+If the encapsulated type is itself a `SProp`, it can be eliminated. This is again a subsingleton elimination because there is a primitive syntactic class of subsingleton types having this property by default:
+  - the empty inductive type (possibly with indices)
+  - any type previously declared in `SProp` (e.g. `SSquash A`)
+  - any record of such types, as in the record presentation of `True` or `and`, without recursivity (and since a record, without indices)
+
+In the paper "Definitional Proof-Irrelevance without K", this class is extended to:
+  - non-record sigma-types of types in `SProp`
+  - inductive types with disjoint indices (sections 5 and 6 of the paper), a priori to be taken in small types, i.e. hsets
+
+As above, if the encapsulated type is not a syntactic `SProp` but only equivalent to a syntactic `SProp`, it can nevertheless be eliminated by using a proxy. For instance, non-record sigma-types of types in `SProp` can be eliminated by first mapping them to their record alternative. Similarly, inductive types with disjoint indices can be eliminated by mapping them to their definition as an `SProp` defined by recursion on the indices. Avoiding a proxy in these cases is the purpose of points 1 and 3 below.
+
+Proofs of types of `SProp` are definitionally undistinguishable and this prevents extending `SProp` with types for which proofs matter (such as `Acc` whose proofs matter for decidability, or types in `HProp` for which belonging to `HProp` is undecidable).
+
+We may want `match` on arbitrary elements of inhabited `SProp` to automatically reduce. This is the question of implementing eta-expansion for positive types and it is addressed elsewhere.
+
+### Late detection of `SProp`
+
+Actual `SProp` has no template polymorphism. Its subsingleton elimination is restricted to `False` so that dynamic computation of subsingleton elimination does not matter.
+
+We do not recommend template polymorphism for `SProp` (since template polymorphism is already planned to be superseded with universe polymorphism).
+
+In particular, we do not plan late detection of a parametric type in `Type` as an `SProp` liable to be considered impredicative. Late detection of a parametric type in `SProp` as a type supporting subsingleton elimination is the purpose of point 2 (once points 1 and 3 are achieved).
+
+## Subtyping of `SProp` in `Prop`
```

Comment:
This will never happen in Coq (except as an unsafe option)
I'm not convinced that it's decidable, and even if it is, deciding a conversion `a == b : A` would need to do some check that there is no `B : SProp` such that `a : B` and `b : B` which sounds quite costly.

Reply:
Wouldn't the mark computed for `a` and `b` be enough to know that `a` and `b` are in some `SProp` type `B`, independently on `A`?

Or are you talking about the cost of computing the mark for `a sTrue` when `a` is `forall A:Type, A`?

I probably don't have the whole picture in mind as much as you do, but at worst, I suspect that we could do it if the internal representation of terms kept track of subtyping, since that would not be different in fine, except that it is made automatically, than asking the user to add explicit `Box` coercions?

Reply:
Marks only work when they're stable by substitution, subtyping breaks that.

Reply:
For the reason above right? That if `a` is of type `forall A:Type, A`, it assumes that the sort of `A` is stable by substitution.